keps/sig-api-machinery/2211-omit-managed-fields-server-side/README.md

```diff
@@ -0,0 +1,372 @@
+
+const (
+    ContentTypeJSON     string = "application/json"
+    ContentTypeYAML     string = "application/yaml"
+    ContentTypeProtobuf string = "application/vnd.kubernetes.protobuf"
+)
+```
+
+2. Add a new serializer to `k8s.io/apimachinery/pkg/runtime/serializer/` which wraps the existing serializers and
+omits the managedFields before encoding:
+```go
+type serializer struct {
+    inner      runtime.Serializer
+}
+
+func NewSerializer(s runtime.Serializer) runtime.Serializer {
+    if s == nil {
+        return nil
+    }
+    return &serializer{inner: s}
+}
+
+func (s *serializer) Encode(obj runtime.Object, w io.Writer) error {
+    if _, err := meta.Accessor(obj); err == nil {
+        obj = obj.DeepCopyObject()
+        a, _ := meta.Accessor(obj)
+        a.SetManagedFields(nil)
+    } else if meta.IsListType(obj) {
+        obj = obj.DeepCopyObject()
+        _ = meta.EachListItem(obj, func(item runtime.Object) error {
+            a, err := meta.Accessor(item)
+            if err != nil {
+                // ignore
+                return nil
+            }
+            a.SetManagedFields(nil)
+            return nil
+        })
+    }
+    return s.inner.Encode(obj, w)
+}
+
+func (s *serializer) Decode(data []byte, defaults *schema.GroupVersionKind, into runtime.Object) (runtime.Object, *schema.GroupVersionKind, error) {
+    return s.inner.Decode(data, defaults, into)
+}
+```
+
+3. Make `k8s.io/apimachinery/pkg/runtime/serializer.newSerializersForScheme` returns serializers for the new "unmanaged" content types.
+For instance:
+```go
+import (
+    ...
+    "k8s.io/apimachinery/pkg/runtime/serializer/unmanaged"
+)
+
+func newSerializersForScheme(scheme *runtime.Scheme, mf json.MetaFactory, options CodecFactoryOptions) []serializerType {
+    ...
+    serializers := []serializerType{
+        jsonSerializerType,
+        {
+            AcceptContentTypes: []string{runtime.ContentTypeYAML},
+            ContentType:        runtime.ContentTypeYAML,
+            FileExtensions:     []string{"yaml"},
+            EncodesAsText:      true,
+            Serializer:         yamlSerializer,
+        },
+        {
+            AcceptContentTypes: []string{runtime.ContentTypeProtobuf},
+            ContentType:        runtime.ContentTypeProtobuf,
+            FileExtensions:     []string{"pb"},
+            Serializer:         protoSerializer,
+
+            Framer:           protobuf.LengthDelimitedFramer,
+            StreamSerializer: protoRawSerializer,
+        },
+    }
+    ...
+
+    allSerializers := append([]serializerType{}, serializers...)
+    for _, serializer := range serializers {
+        serializer.ContentType += runtime.ContentTypeSuffixOmitManagedFields
+        for i := range serializer.AcceptContentTypes {
+            serializer.AcceptContentTypes[i] += runtime.ContentTypeSuffixOmitManagedFields
+        }
+        serializer.Serializer = unmanaged.NewSerializer(serializer.Serializer)
```

Comment:
Let's not use the word "unmanaged" at all. We are not producing unmanaged objects.